src/v/storage/segment_utils.cc

```diff
@@ -700,18 +694,51 @@ ss::future<compaction_result> self_compact_segment(
         // will survive until it's used in self compaction below.
```

Comment:
This comment is outdated. "self compaction below" is not a thing after refactoring.

Spent 20 minutes trying to reason about the flow of the lock holder. Would be useful to have a more detailed comment about why we're passing the lock holder around the way we do.

Reply:
Good catch, thank you.

I think it's easiest to understand the flow of the lock holder by looking at the old version of this code: (https://github.com/redpanda-data/redpanda/blob/d96b1226382270c95de1cc96029acc3d47c9aba0/src/v/storage/segment_utils.cc#L671).

This new function is called in two places, once in `self_compact_segment()` and once in `build_offset_map()`. 

In `self_compact_segment()`, we care about persisting our hold on the lock, as denoted by the variable name `read_holder`. To make this a non-functional change and not alter how we do anything with the locking, I have a `ss::rwlock::holder&` input parameter to `maybe_rebuild_compaction_index()`, so the acquisition/life-time of locks there doesn't change.

In `build_offset_map()`, we don't care about persisting the lock after rebuilding the compaction index, so the variable is denoted `read_lock` and allowed to go out of scope after `maybe_rebuild...()` is called.

Reply:
Added more to the comment in the header here, let me know if you think that description makes the code content a bit more digestible.